### PR TITLE
Raise more verbose warning for gradient scaling in scheduler/optimizer case

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes https://github.com/Lightning-AI/pytorch-lightning/issues/21339
Fixes https://github.com/Lightning-AI/pytorch-lightning/issues/5558

This is both a new and old issue. When gradient scaling is applied, `optimizer.step()` may be skipped if `nan` or `inf` is detected, and the learning rate scheduler detects that the optimizer didn’t step, which results in the warning that users are seeing. There is really nothing we can do to solve this, it requires some effort on the Pytorch side:
https://github.com/pytorch/pytorch/issues/67590
but it has no activity for 4 years. This PR is an attempt at at least provide a bit more verbose output to the end user. 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--21426.org.readthedocs.build/en/21426/

### CODE REVIEW

File: src/lightning/pytorch/core/module.py [hunk 1]
```diff
@@ -1329,10 +1330,26 @@ def lr_scheduler_step(self, scheduler, metric):
...
+
+            if metric is None:
+                scheduler.step()  # type: ignore[call-arg]
+            else:
+                scheduler.step(metric)
+
+        for w in caught:
+            msg = str(w.message)
+
+            if "lr_scheduler.step()" in msg and "optimizer.step()" in msg:
```
Comment (GdoongMathew) [inline src/lightning/pytorch/core/module.py:1344]:
```suggestion
            if "Detected call of `lr_scheduler.step()` before `optimizer.step()`." in msg:
```

There's another warning ([here](https://github.com/pytorch/pytorch/blob/main/torch/optim/lr_scheduler.py#L258-L261)) that fit in the original if statement. The proposed one should be more specific.
